econml/sklearn_extensions/selective_regularization.py

```diff
@@ -0,0 +1,107 @@
+
+    def fit(self, X, y, sample_weight=None):
+        self._penalized_inds = np.delete(np.arange(X.shape[1]), self._unpenalized_inds)
+        X1 = X[:, self._unpenalized_inds]
+        X2 = X[:, self._penalized_inds]
+
+        X2_res = X2 - LinearRegression(fit_intercept=self._fit_intercept).fit(X1, X2,
+                                                                              sample_weight=sample_weight).predict(X1)
+        y_res = y - LinearRegression(fit_intercept=self._fit_intercept).fit(X1, y,
+                                                                            sample_weight=sample_weight).predict(X1)
+
+        if sample_weight is not None:
+            self._penalized_model.fit(X2_res, y_res, sample_weight=sample_weight)
+        else:
+            self._penalized_model.fit(X2_res, y_res)
+
+        # The unpenalized model can't contain an intercept, because in the analysis above
+        # we rely on the fact that M(X beta) = (M X) beta, but M(X beta + c) is not the same
+        # as (M X) beta + c, so the learned coef and intercept will be wrong
+        intercept = self._penalized_model.predict(np.zeros_like(X2[0:1]))
+        if not np.allclose(intercept, 0):
+            raise AttributeError("The penalized model has a non-zero intercept; to fit an intercept "
+                                 "you should instead either set fit_intercept to True when initializing the "
+                                 "SelectiveRegression instance (for an unpenalized intercept) or "
+                                 "explicitly add a column of ones to the data being fit and include that "
+                                 "column in the penalized indices.")
+
+        # now regress X1 on y - X2 * beta2 to learn beta1
+        self._model_X1 = LinearRegression(fit_intercept=self._fit_intercept)
+        self._model_X1.fit(X1, y - self._penalized_model.predict(X2), sample_weight=sample_weight)
+
+        return self
+
+    def predict(self, X):
+        X1 = X[:, self._unpenalized_inds]
+        X2 = X[:, self._penalized_inds]
+        return self._model_X1.predict(X1) + self._penalized_model.predict(X2)
+
+    @property
+    def coef_(self):
+        y_dim = self._model_X1.coef_.shape[0:-1]  # get () if y was a vector, or (d_y,) otherwise
+        x_dim = len(self._penalized_inds) + len(self._unpenalized_inds),
+        coefs = np.empty(y_dim + x_dim)
+        coefs[..., self._penalized_inds] = self._penalized_model.coef_
+        coefs[..., self._unpenalized_inds] = self._model_X1.coef_
+        return coefs
+
+    @property
+    def intercept_(self):
+        if not self._fit_intercept:
+            raise AttributeError("intercept_ property is only supported when fit_intercept=True")
+        # Note that the penalized model should *not* have an intercept
+        return self._model_X1.intercept_
+
+    def score(self, X, y):
+        """Score the predictions for a set of features to ground truth."""
```

Comment:
Similarly requires check_fitted and maybe check_inputs.

Reply:
Done.